The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12257213480/job/34194196401) [(plain)](https://github.com/rust-lang/rust/commit/a5914c823b93be5b9012f2a5e931ff54baff074a/checks/34194196401/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling std v0.0.0 (/checkout/library/std)
thread 'rustc' panicked at compiler/rustc_interface/src/queries.rs:36:25:
already borrowed: BorrowMutError
stack backtrace:
   0:     0x7f0f9b1a9141 - <<std[f727eb1ed3cbfa49]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[8e2c8c99b037a10c]::fmt::Display>::fmt
   1:     0x7f0f9b206183 - core[8e2c8c99b037a10c]::fmt::write
   2:     0x7f0f9b19cb09 - <std[f727eb1ed3cbfa49]::sys::pal::unix::stdio::Stderr as std[f727eb1ed3cbfa49]::io::Write>::write_fmt
   3:     0x7f0f9b1a8fe2 - <std[f727eb1ed3cbfa49]::sys::backtrace::BacktraceLock>::print
   4:     0x7f0f9b1aba50 - std[f727eb1ed3cbfa49]::panicking::default_hook::{closure#1}
   5:     0x7f0f9b1ab7cf - std[f727eb1ed3cbfa49]::panicking::default_hook
   6:     0x7f0f968ae488 - <alloc[9ba8a48ea8c59ee]::boxed::Box<rustc_driver_impl[fec526931b25148d]::install_ice_hook::{closure#0}> as core[8e2c8c99b037a10c]::ops::function::Fn<(&dyn for<'a, 'b> core[8e2c8c99b037a10c]::ops::function::Fn<(&'a std[f727eb1ed3cbfa49]::panic::PanicHookInfo<'b>,), Output = ()> + core[8e2c8c99b037a10c]::marker::Sync + core[8e2c8c99b037a10c]::marker::Send, &std[f727eb1ed3cbfa49]::panic::PanicHookInfo)>>::call
   7:     0x7f0f9b1ac3c8 - std[f727eb1ed3cbfa49]::panicking::rust_panic_with_hook
   8:     0x7f0f9b1abfde - std[f727eb1ed3cbfa49]::panicking::begin_panic_handler::{closure#0}
   9:     0x7f0f9b1a9769 - std[f727eb1ed3cbfa49]::sys::backtrace::__rust_end_short_backtrace::<std[f727eb1ed3cbfa49]::panicking::begin_panic_handler::{closure#0}, !>
  11:     0x7f0f9683e5d0 - core[8e2c8c99b037a10c]::panicking::panic_fmt
  12:     0x7f0f9683e4c3 - core[8e2c8c99b037a10c]::cell::panic_already_borrowed
  13:     0x7f0f96cc61ab - <rustc_interface[c0f98b4a6dd3f783]::queries::Queries>::parse
  14:     0x7f0f96cc6294 - <rustc_interface[c0f98b4a6dd3f783]::queries::Queries>::global_ctxt
---
warning: the ICE couldn't be written to `/checkout/rustc-ice-2024-12-10T13_41_27-11378.txt`: Read-only file system (os error 30)

note: rustc 1.85.0-nightly (9c445a0d6 2024-12-10) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C embed-bitcode=yes -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
end of query stack
error: could not compile `core` (lib)

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc /checkout/obj/build/bootstrap/debug/rustc --crate-name core --edition=2021 library/core/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 --warn=unexpected_cfgs --check-cfg 'cfg(bootstrap)' --check-cfg 'cfg(no_fp_fmt_parse)' --check-cfg 'cfg(stdarch_intel_sde)' --check-cfg 'cfg(target_arch, values("xtensa"))' --check-cfg 'cfg(feature, values(any()))' -C debug-assertions=on --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values("debug_refcell", "optimize_for_size", "panic_immediate_abort"))' -C metadata=c9690195bf6f5d1b -C extra-filename=-c9690195bf6f5d1b --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps -Csymbol-mangling-version=v0 '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap)' -Zmacro-backtrace -Csplit-debuginfo=off -Cprefer-dynamic -Zinline-mir -Zinline-mir-preserve-debug -Clink-args=-Wl,-z,origin '-Clink-args=-Wl,-rpath,$ORIGIN/../lib' -Cembed-bitcode=yes -Cforce-frame-pointers=yes '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' -Z binary-dep-depinfo` (exit status: 101)
Build completed unsuccessfully in 0:05:08
  local time: Tue Dec 10 13:41:27 UTC 2024
  network time: Tue, 10 Dec 2024 13:41:27 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/compiler/rustc_driver_impl/src/lib.rs
+++ b/compiler/rustc_driver_impl/src/lib.rs
@@ -173,7 +173,7 @@ pub trait Callbacks {
     fn after_expansion<'tcx>(
         &mut self,
         _compiler: &interface::Compiler,
-        _queries: &'tcx Queries<'tcx>,
+        _tcx: TyCtxt<'tcx>,
     ) -> Compilation {
         Compilation::Continue
     }
@@ -425,18 +425,18 @@ fn run_compiler(
                 return early_exit();
             }
 
-            // Make sure name resolution and macro expansion is run.
-            queries.global_ctxt().enter(|tcx| tcx.resolver_for_lowering());
-
-            if let Some(metrics_dir) = &sess.opts.unstable_opts.metrics_dir {
-                queries.global_ctxt().enter(|tcxt| dump_feature_usage_metrics(tcxt, metrics_dir));
-            }
-
-            if callbacks.after_expansion(compiler, queries) == Compilation::Stop {
-                return early_exit();
-            }
-
             queries.global_ctxt().enter(|tcx| {
+                // Make sure name resolution and macro expansion is run.
+                let _ = tcx.resolver_for_lowering();
+
+                if let Some(metrics_dir) = &sess.opts.unstable_opts.metrics_dir {
+                    dump_feature_usage_metrics(tcx, metrics_dir);
+                }
+
+                if callbacks.after_expansion(compiler, tcx) == Compilation::Stop {
+                    return early_exit();
+                }
+
                 passes::write_dep_info(tcx);
 
                 if sess.opts.output_types.contains_key(&OutputType::DepInfo)

--- a/compiler/rustc_driver_impl/src/lib.rs
+++ b/compiler/rustc_driver_impl/src/lib.rs
@@ -45,7 +45,7 @@ use rustc_errors::registry::Registry;
 use rustc_errors::{ColorConfig, DiagCtxt, ErrCode, FatalError, PResult, markdown};
 use rustc_feature::find_gated_cfg;
 use rustc_interface::util::{self, get_codegen_backend};
-use rustc_interface::{Linker, Queries, interface, passes};
+use rustc_interface::{Linker, interface, passes};
 use rustc_lint::unerased_lint_store;
 use rustc_metadata::creader::MetadataLoader;
 use rustc_metadata::locator;
@@ -158,13 +158,10 @@ pub trait Callbacks {
     /// Called after parsing the crate root. Submodules are not yet parsed when
     /// this callback is called. Return value instructs the compiler whether to
     /// continue the compilation afterwards (defaults to `Compilation::Continue`)
-    #[deprecated = "This callback will likely be removed or stop giving access \
-                    to the TyCtxt in the future. Use either the after_expansion \
-                    or the after_analysis callback instead."]
-    fn after_crate_root_parsing<'tcx>(
+    fn after_crate_root_parsing(
         &mut self,
         _compiler: &interface::Compiler,
-        _queries: &'tcx Queries<'tcx>,
+        _queries: &ast::Crate,
     ) -> Compilation {
         Compilation::Continue
     }
@@ -416,8 +413,9 @@ fn run_compiler(
                 return early_exit();
             }
 
-            #[allow(deprecated)]
-            if callbacks.after_crate_root_parsing(compiler, queries) == Compilation::Stop {
+            if callbacks.after_crate_root_parsing(compiler, &*queries.parse().borrow())
+                == Compilation::Stop
+            {
                 return early_exit();
             }
 

--- a/compiler/rustc_middle/src/query/plumbing.rs
+++ b/compiler/rustc_middle/src/query/plumbing.rs
@@ -44,16 +44,16 @@ pub struct DynamicQuery<'tcx, C: QueryCache> {
     pub format_value: fn(&C::Value) -> String,
 }
 
-pub struct QuerySystemFns<'tcx> {
+pub struct QuerySystemFns {
     pub engine: QueryEngine,
     pub local_providers: Providers,
     pub extern_providers: ExternProviders,
-    pub encode_query_results: fn(
+    pub encode_query_results: for<'tcx> fn(
         tcx: TyCtxt<'tcx>,
         encoder: &mut CacheEncoder<'_, 'tcx>,
         query_result_index: &mut EncodedDepNodeIndex,
     ),
-    pub try_mark_green: fn(tcx: TyCtxt<'tcx>, dep_node: &dep_graph::DepNode) -> bool,
+    pub try_mark_green: for<'tcx> fn(tcx: TyCtxt<'tcx>, dep_node: &dep_graph::DepNode) -> bool,
 }
 
 pub struct QuerySystem<'tcx> {
@@ -68,7 +68,7 @@ pub struct QuerySystem<'tcx> {
     /// This is `None` if we are not incremental compilation mode
     pub on_disk_cache: Option<OnDiskCache>,
 
-    pub fns: QuerySystemFns<'tcx>,
+    pub fns: QuerySystemFns,
 
     pub jobs: AtomicU64,
 }

